### PR TITLE
Pass validation_context to validates_associated

### PR DESCRIPTION
**Pull request by austenmadden**
### Motivation / Background

This Pull Request has been created because https://github.com/rails/rails/issues/46239

### Detail

The docs imply that when specified the `on` option of `validates_associated` should be respected by the validated association. Prior to this change the validation context used will simply be the default behavior as if no `on` option were specified. With this change, the validation_context of the parent record is passed to the `valid?` check.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] There are no typos in commit messages and comments.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Feature branch is up-to-date with `main` (if not - rebase it).
* [x] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [x] Tests are added if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] PR is not in a draft state.
* [x] CI is passing.

### CODE REVIEW

**Comment by skipkayhil** [comment]
Hey @austenmadden, thanks for opening a PR! I was thinking about some different scenarios and I can't decide whether I think this is expected behavior or not:

```
def test_validates_associated_with_create_context
    Reply.validates_associated :topic, on: :create
    Topic.validates_presence_of :content, on: :create
    t = Topic.create!(title: "uhoh", content: "stuff")
    t.update!(content: nil)
    t.replies.create!("title" => "A reply", "content" => "with content!") # => what should happen here?
end
```

Should creating an associated record trigger validations for the `:update` context on Topic or the `:create` context? My gut instinct says `:create` would be unexpected

**Comment by austenmadden** [comment]
@skipkayhil Hey thanks for following up! I see your point. However, I feel like we should do something about the status quo as it's definitely makes `validates_associated` tricky with custom contexts IMO. 

Here's an example where the current behavior can be kinda wacky without custom contexts too.

```ruby
  def test_validates_associated_with_different_context
    Reply.validates_associated :topic, on: :update
    Topic.validates_presence_of :content, on: :create

    t = Topic.create!(title: "uhoh", content: "stuff")
    t.update_column(:content, nil)
    r = Reply.create!(title: "A reply", topic: t) # Explodes
  end

```

What's interesting however is this snippet doesn't even hit the code being touched here according to the  backtrace/breakpoints 🤔

**Comment by austenmadden** [comment]
@skipkayhil could a compromise be to treat `:create` and `:update` as special cases where we avoid passing the context from the parent?

**Comment by p8** [comment]
We do treat `:create` and `:update` as a special case in autosave as well:
https://github.com/rails/rails/blob/main/activerecord/lib/active_record/autosave_association.rb#L513-L515
Maybe that method belongs on https://github.com/rails/rails/blob/main/activerecord/lib/active_record/validations.rb

**Comment by austenmadden** [comment]
@skipkayhil @p8 If I refactor this PR to treat the `create` and `update` as a special case (and potentially move `custom_validation_context?` to `validations.rb`) do we feel like that's a big enough improvement on the status quo and sidesteps the caveats to this implementation enough to move forward?

**Comment by austenmadden** [comment]
Let me know what you think of the latest commit.

File: activerecord/lib/active_record/validations.rb [hunk 1]
```diff
@@ -73,6 +73,10 @@ def valid?(context = nil)
 
     alias_method :validate, :valid?
 
+    def custom_validation_context?
```
[the remark this answers is not in the document]
Comment (austenmadden) [reply activerecord/lib/active_record/validations.rb]: Thanks!

**Comment by skipkayhil** [comment]
@zzak I think this is related to #24135

**Comment by gstokkink** [comment]
@ghiculescu can this move forward? Fixes #24135 as well I think.

**Comment by ghiculescu** [comment]
It'll need review from someone in the core team, sorry. They'll get to it!

In the meantime could you fix the merge conflicts?

**Comment by austenmadden** [comment]
I'd be happy to!

**Comment by rafaelfranca** [comment]
Duplicated of #24135

**Comment by rafaelfranca** [comment]
This PR is accepted but I'd like to give credits to @rafal-brize for the first try. Can you make him co-author of this commit and add him to the CHANGELOG entry?

**Comment by danielolivaresd** [comment]
@austenmadden friendly reminder about this PR

**Comment by austenmadden** [comment]
Ah sorry y'all, I missed the last couple comments but yes I can do that.

**Pull request merged**